src/server/pkg/worker/api_server.go

```diff
@@ -191,42 +285,51 @@ func HashDatum(data []*pfs.FileInfo, options *Options) (string, error) {
 
 // Process processes a datum.
 func (a *APIServer) Process(ctx context.Context, req *ProcessRequest) (resp *ProcessResponse, retErr error) {
-	defer func(start time.Time) { a.Log(req, resp, retErr, time.Since(start)) }(time.Now())
 	// We cannot run more than one user process at once; otherwise they'd be
 	// writing to the same output directory. Acquire lock to make sure only one
 	// user process runs at a time.
 	a.Lock()
 	defer a.Unlock()
+	logger := a.getTaggedLogger(req)
+	logger.Logf("Received request")
 
-	// ppsserver sorts inputs by input name, so this is stable even if
-	// a.options.Inputs are reordered by the user
-	tag, err := HashDatum(req.Data, a.options)
+	// Hash inputs and check if output is in s3 already. Note: ppsserver sorts
+	// inputs by input name for both jobs and pipelines, so this hash is stable
+	// even if a.Inputs are reordered by the user
+	tag, err := a.HashDatum(req.Data)
 	if err != nil {
 		return nil, err
 	}
 	if _, err := a.pachClient.InspectTag(ctx, &pfs.Tag{tag}); err == nil {
 		// We've already computed the output for these inputs. Return immediately
+		logger.Logf("skipping input, as it's already been processed")
 		return &ProcessResponse{
 			Tag: &pfs.Tag{tag},
 		}, nil
 	}
 
+	// Download input data
+	logger.Logf("input has not been processed, downloading data")
 	if err := a.downloadData(ctx, req.Data); err != nil {
 		return nil, err
 	}
-	// Create output directory (currently /pfs/out)
-	if err := os.MkdirAll(client.PPSOutputPath, 0666); err != nil {
-		return nil, err
-	}
 	defer func() {
 		if err := os.RemoveAll(client.PPSInputPrefix); retErr == nil && err != nil {
 			retErr = err
 			return
 		}
 	}()
-	if log, err := a.runUserCode(ctx); err != nil {
+
+	// Create output directory (currently /pfs/out) and run user code
+	if err := os.MkdirAll(client.PPSOutputPath, 0666); err != nil {
+		return nil, err
+	}
+	logger.Logf("beginning to process user input")
```

Comment:
These log messages are on purpose. The goal was for users to be able to see that pfs saw a datum and skipped it because it had already been processed (vs did not see a datum because it actually isn't in PFS/there's a bug in our code).